### ISSUE TITLE
Unnecessary nullability warnings from `as` expressions

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#34638]
There should be no warnings in the following:
```C#
class Program
{
    static void F1(object x)
    {
        if ((x as object[]) != null)
        {
            _ = (x as object[]).Length;
        }
    }

    static void F2(object y)
    {
        if (y is object[])
        {
            _ = (y as object[]).Length;
        }
    }

    static void F3(object? z)
    {
        if ((z as object[]) != null)
        {
            _ = ((object[])z).Length;
        }
    }
}
```
Result:
```
(7,18): warning CS8602: Possible dereference of a null reference.
(15,18): warning CS8602: Possible dereference of a null reference.
(23,18): warning CS8600: Converting null literal or possible null value to non-nullable type.
(23,18): warning CS8602: Possible dereference of a null reference.
```

**Comment by danmosemsft** [comment]
fyi @safern @krwq @buyaa-n

**Comment by gafter** [comment]
We currently specify that as-expressions are always possibly null unless (1) the left-hand-side is not null, and (2) the conversion to the as-type is an implicit reference conversion.  This is therefore by design.

**Comment by safern** [comment]
But if we're checking with an `is` or a `!= null` in a previous if/debug.assert, shouldn't the compiler infer that the following cast will never return null?

**Comment by danmosemsft** [comment]
cc @stephentoub

**Comment by gafter** [comment]
By design the state that we track for a variable is its static type, and whether or not it is believed to be non-null.  We don't track information like what type it is believed to be convertible to.

**Comment by danmosemsft** [comment]
If this was a sufficiently common pattern (don't know), would it be potentially reasonable to do that? Or technically infeasible?

**Comment by jaredpar** [comment]
@danmosemsft 

> Or technically infeasible?

For identifiers like locals this would be technically feasible. The tracking comes down to essentially two types of items:

1. The set of types we know a local is convertible to: one concrete type and a collection of interfaces.
1. Tracking every potential write to a local and invalidating that list as appropriate based on the type of the expression being assigned. 

That's a bit of extra state tracking there though. Same logic could be applied to field access off of locals too. 

For properties though it's a bit dicier. Yes we could continue to track it but it's bit more questionable if we should. Today NRT operates off the assumption that if a property invocation is non-null once, it will remain non-null in the future. Extending the `as` check to properties would extend that assumption to saying it's non-null and continues to have the same runtime type. Not sure if we want to do that.

> If this was a sufficiently common pattern (don't know), would it be potentially reasonable to do that?

My push back here would be to move to more modern patterns that are both more efficient (avoid the double cast) and work naturally with NRT. Consider the small rewrite of the code above that has no NRT warnings.

```
#nullable enable
class Program
{
    static void F1(object x)
    {
        if (x is object[] y)
        {
            _ = y.Length;
        }
    }

    static void F2(object y)
    {
        if (y is object[] x)
        {
            _ = x.Length;
        }
    }

    static void F3(object? z)
    {
        if (z is object[] y)
        {
            _ = y.Length;
        }
    }
}
```

Closing for now as this is the designed behavior. If we want to push on this though let me know and we can open an issue in csharplang.

**Comment by stephentoub** [comment]
> My push back here would be to move to more modern patterns

In general, that's the right answer.  Right now we're trying to annotate with as a few modifications as possible that actually impact the generated IL, because in theory any such change can impact perf, and we don't have the time right now to benchmark every such possible change.

**Comment by danmosemsft** [comment]
@stephentoub is it possible for the specific translation above to impact perf?

**Comment by stephentoub** [comment]
> is it possible for the specific translation above to impact perf?

If anything it would likely help, if it were measurable at all.

**Comment by danmosemsft** [comment]
Right, so I was thinking its reasonable to make this particular refactoring as we go. (I was confused as you mentioned perf)

**Comment by stephentoub** [comment]
> Right, so I was thinking its reasonable to make this particular refactoring as we go. (I was confused as you mentioned perf)

Probably... I just don't like the idea of making such changes without verification, at least on code paths that are potentially impactful... don't know if this one is.
